integration/oasis/stationfinder/stations_iterator_alg.go

```diff
@@ -129,3 +131,96 @@ func buildChargeStationInfoDict(iter nearbyStationsIterator) map[string]bool {
 
 	return dict
 }
+
+type WeightBetweenNeighbors struct {
+	c   []CostBetweenChargeStations
+	err error
+}
+
+// CalculateWeightBetweenNeighbors accepts locations array, which will search for nearby
+// charge stations and then calculate weight between stations, the result is used to
+// construct graph.
+// - The input of locations contains: orig location -> first place to search for charge ->
+//   second location to search for charge -> ... -> dest location
+// - Both search nearby charge stations and calculate weight between stations are heavy
+//   operations, so put them into go-routine and use waitgroup to guarantee result channel
+//   is closed after everything is done.
+// - CalcWeightBetweenChargeStationsPair needs two iterators, one for nearbystationiterator
+//   represents from location and one for next location.  An array of channel is created
+//   to represent whether specific iterator is ready or not.
+// - All iterators has been recorded in iterators array
+//   @Todo: isIteratorReady could be removed later.  When iterator is not ready, should
+//         pause inside iterator itself.  That need refactor the design of stationfinder.
+func CalculateWeightBetweenNeighbors(locations []*StationCoordinate, oc *osrmconnector.OSRMConnector, sc *searchconnector.TNSearchConnector) chan WeightBetweenNeighbors {
+	c := make(chan WeightBetweenNeighbors)
+
+	if len(locations) > 2 {
+		iterators := make([]nearbyStationsIterator, len(locations))
+		isIteratorReady := make([]chan bool, len(locations))
+		for i := range isIteratorReady {
+			isIteratorReady[i] = make(chan bool)
+		}
+		var wg sync.WaitGroup
+
+		for i := 0; i < len(locations); i++ {
+			if i == 0 {
+				wg.Add(1)
+				go func(wg *sync.WaitGroup, first int) {
+					iterators[first] = NewOrigIter(locations[first])
+					isIteratorReady[first] <- true
+					wg.Done()
+					glog.Info("Finish generating NewOrigIter")
+				}(&wg, i)
```

Comment:
capture from context maybe good enough?

Reply:
I think pass `i` is good, but I'll suggest to capture `wg` instead of pass it. 
There's no `reference` in `golang`, either pass by value or pass by pointer. In this context, capture `wg` I think is capture the var itself, so don't need to consider whether its value or pointer.

Reply:
`waitgroup` is the one I met error before, when write a function I passed `wg sync.WaitGroup` but which will causing `wg.Wait()` hang there forever.

I will make small experiment to verify capture, if `func()` could capture which correctly, will only keep `i`

Reply:
I think go routine could capture it correctly.  https://play.golang.org/p/gZ2-UZ-2NTj

Will remove this parameter.